### PR TITLE
Documentation ( Installation, Quick Start), Changelog template , Release process

### PR DESCRIPTION
**Pull request by lalitb**
-  Updated README.md with notes on Installation and Quick Start.
-  Added  Changelog.md template file.
-  Added Release process. ( Similar to how it is done across other open-telemetry projects )

To summarize the release process as part of this PR:

1. CHANGELOG.md is managed manually, and as part of raising any PR, dev needs to add the summary of change under `Unreleased` section.
2. On day of release, Release manager decides on a tag say ( v0.0.1-beta).
3. He/She runs pre_release.sh which
     - creates a branch ( pre_release_v0.0.1-beta) 
     - modifies CHANGELOG.md ( Search for `##Unreleased` section and adds a new section with `<tag> - <date>`  header beneath it. That way `##Unreleased` section remains intact for new changes, and existing change-notes go within newly created section. I will reword the documentation.) 
4. He/She creates a PR for these changelog.md changes, and waits for it to be merged to master.
5.  Once PR is merged, he creates the above decided tag on the new merged commit hash.
6. He/She then goes to the github release UI page, and select "Create New Release". Specifies the tag created in step 5, and takes the curated contents as release notes, and creates the release.

### CODE REVIEW

File: pre_release.sh
Comment (ThomsonTan) [inline pre_release.sh:19]: nit: give some example on semver in comment?

File: pre_release.sh
Comment (ThomsonTan) [inline pre_release.sh:34]: Does this mean `git diff --quiet` exists with `0` if there is any local change?

File: pre_release.sh
Comment (reyang) [inline pre_release.sh:59]:
```suggestion
# update CHANGELOG.md
```

**Review by reyang — APPROVED** [review]
LGTM with some minor formatting suggestions.

File: pre_release.sh
Comment (ThomsonTan) [inline pre_release.sh:35]: nit: echo "Error: ..."?

File: pre_release.sh
Comment (maxgolov) [inline pre_release.sh:42]: How does manually managing `CHANGELOG.md` interact with built-in GitHub release management tab? Is intention to populate the markdown, then copy individual sections into Release tab?

File: README.md
Comment (maxgolov) [inline README.md:23]: I think we need to decide on a few practical things. When you create a package, what is the actual installation / deployment layout:

- /usr/local/lib
- /usr/local/include/opentelemetry - for API?
- /usr/lcoal/include/opentelemetry/sdk - fo SDK developers?

# Why this needs to be described

Many packaging tools, e.g. `vcpkg` or `deb` would prefer you to have a clear-cut definition of what are the ship'able artifacts, even if include headers. Usually the layout for C++ projects requires a single collated directory. In our case it is two directories: one for customers `api` and another for devs, `sdk`... Are we shipping two separate packages, such as `opentelemetry` and `libopentelemetry-dev` - for developers?

File: README.md
Comment (maxgolov) [inline README.md:37]: This may be a bit too generic. We need more detailed step-by-step instructions here, such as min cmake version, min compiler version, min required versions of Google Test. Just having generic `cmake` instruction is a good start. But may not necessarily be sufficient. For `ctest` - we can run this if built with tests as part of a single `cmake build` command.

File: RELEASING.md
Comment (maxgolov) [inline RELEASING.md:7]: My biggest concern here, is right now you can easily author a release tag thru GitHub UI. For that - there is also new GH console utility, which can be used to automate all GH UI abilities, including creating a new tag + adding release notes. It may be unnecessary to have a separate script for that, if we say that we use the GitHub release process, documented here:

https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/managing-releases-in-a-repository

That way even a person without console, e.g. Program Manager or Project Manager, can efficiently work on creating a new release quickly, without running any special tooling. I think it's a big advantage.

File: pre_release.sh
Comment (maxgolov) [reply pre_release.sh:19]: I think we're a little bit reinventing the wheel here, since most of this - including `semver` convention is documented here: https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/managing-releases-in-a-repository - why not adopt this process?

**Comment by maxgolov** [comment]
For managing release process - I'd suggest we use:
- GitHub UX for Release management: it covers creating tags in semver and creating release notes.
- for automated (nightly) - use the latest technology available: https://cli.github.com/ - that is `gh release create`: https://cli.github.com/manual/gh_release_create - which is gonna be harder for us to create something better than this in a limited time given.

Both options (GitHub Release tab and `gh` CLI) are industry standard. It is easy for any person, not necessarily dev, to use these tools.

File: README.md
Comment (maxgolov) [inline README.md:42]: To elaborate a little bit on this, ideally - the local dev build process could be as easy as:

- `tools/build.cmd` - on Windows, which will auto-detect the compiler AND setup all the build tools needed (unless you want to explicitly specify the compiler, for that we'll have `tools/build-vs201x.cmd` )

- `tools/build.sh` - for anything POSIX that only has `/bin/bash` : Linux, Mac, Android.

And

- crossplat `vcpkg install opentelemetry[:feature_pack]`

Maybe we can remove the cmake section here? And I can elaborate on a separate doc with the tooling I'm adding, as t's near-zero effort to onboard. I already added some drafts for `vcpkg` build process in the other PR.

Comment (lalitb) [reply README.md:42]: sure, I can remove the steps from here, and instead add one-liner saying Refer to (say) INSTALL.md for details.

File: RELEASING.md
Comment (lalitb) [reply RELEASING.md:7]: The  github UI is supposed to be used to publish a release as mentioned in final section of this document  (https://github.com/open-telemetry/opentelemetry-cpp/blob/c0ea1fd587e1be0d99a97d639a1f82994515000c/RELEASING.md#release). I can make it explicit if this is not clear in document. The idea behind the first two steps are to manage the changelog.md. I am fine removing the second step ( Tagging ), and let the tag be created automatically by github UI if that's the ask ?

**Comment by lalitb** [comment]
> For managing release process - I'd suggest we use:
> 
> * [GitHub UX for Release management](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/managing-releases-in-a-repository) - it covers creating tags in semver and creating release notes.
> * for automated (nightly) - use the latest technology available: https://cli.github.com/ - that is `gh release create`: https://cli.github.com/manual/gh_release_create - which is great. It is gonna be harder for us to create something better than this in a limited time given.
> 
> Both options (GitHub Release tab and `gh` CLI) are industry standard. These are designed to be easy-to-use by any person, not necessarily dev. Plus, we'd have less of our own tooling to maintain.

The github UI is supposed to be used to publish a release as mentioned in final section of this document (https://github.com/open-telemetry/opentelemetry-cpp/blob/c0ea1fd587e1be0d99a97d639a1f82994515000c/RELEASING.md#release). I can make it explicit if this is not clear in document. The idea behind the first two steps are to manage the changelog.md. I am fine removing the second step ( Tagging ), and let the tag be created automatically by github UI if that's the ask ?

File: README.md
Comment (lalitb) [reply README.md:37]: Sure, we can reference to "INSTALL.md" document, as you plan to document it separately ?

File: pre_release.sh
Comment (lalitb) [reply pre_release.sh:19]: Please refer to my comment on managing changelog.md.

File: pre_release.sh
Comment (lalitb) [reply pre_release.sh:34]: This is checking the (standard ) output from the command, and not the return code (`$?`) of the command  So if there is any local changes, the command would print those changes. The return code would remain 0 irrespective of any local changes or not.

File: pre_release.sh
Comment (lalitb) [reply pre_release.sh:42]: Correct

File: README.md
Comment (lalitb) [reply README.md:23]: I am now confused here. As per my understanding, and also explicitly specified in this PR, we are only going to have source-only releases in same layout as the repo ( same as provided by github release ui by default). Which would mean, it would be responsibility of developers to decide how they would like to package this and distribute further - which should ideally be the structure as you mentioned above. Do you mean, even though we provide source-only releases, there should be recommendation for developer how to package them ?

Comment (maxgolov) [reply README.md:23]: In order to consume the package with tooling, like `vcpkg` (which builds the release for you from source and installs it for you), we need to identify the layout of how what we expect to get as a result of the build. In other words, if you build from source and invoke `make install` - what is the expected outcome?

File: README.md
Comment (maxgolov) [reply README.md:37]: I can cover this part. You can remove the build instructions for now.

File: RELEASING.md
Comment (maxgolov) [reply RELEASING.md:7]: We can probably skip the auto-tagging, and suggest that Release Manager uses their best judgement to tag the needed release sequentially, using semver 2.0, on that exact branch they wanna tag. We can keep the auto-tagging part if we want to cover tagging nightlies (but I am not sure we want this, as it'll create a ton of tags). One imporant moment where we may need to get the tag of the repo is to generate the `version.h` from the tag, not auto-increment the tag itself.

File: README.md
Comment (maxgolov) [reply README.md:42]: I can post the recommendations in `INSTALL.md`, to cover various  ways to build (or build from source and install) instructions. Separate sections for common build options:
- CMake
- Bazel
- vcpkg
- Visual Studio 2019 with CMake support (GUI build, no command-line tooling)
- command line wrappers to facilitate the build
- docker cross-plat builds

I'll send a separate PR with the doc. Most of it is already implemented in #377

**Comment by maxgolov** [comment]
> let the tag be created automatically by github UI if that's the ask ?

Yeah, we can assume that Release Manager can use their best judgement and UI (no scripts) to generate the next matching tag, given UI totally allows to do this without running any command-line tooling ( or `gh` utility can be used by those who prefer command line tooling).

I think it's more valuable to write the opposite script: generate `version.h` from git tag, so that when the build is built - we properly report the actual SDK version.

File: README.md
Comment (lalitb) [reply README.md:23]: Ok, thanks for clarification. I will add expected deployment layout structure from packaging tools propspective in the document. 
As of now, we are installing api headers at `/usr/local/include/opentelemetry` for cmake builds. We can extended that as suggested above:

```
/usr/local/lib  -> For libraries
/usr/local/include/opentelemetry -> for api headers
/usr/local/include/opentelemetry/sdk -> for sdk headers
```

--
Are we shipping two separate packages, such as opentelemetry and libopentelemetry-dev - for developers?
 --
We are not shipping anything :) Although same recommendation on the final library/header deployment structure would apply for source only package like `libopentelemetry-dev` once it is `installed` and build from `/usr/local/src`.

File: README.md
Comment (lalitb) [reply README.md:37]: I have added reference to INSTALL.md under this section. I still feel it makes sense to keep generic installation steps as it given some idea to devs/users on how to build the library. I have already mentioned "with all dependencies specified". Regarding ctest, the default build using `cmake build` does build the tests , so these command should work out of box to run tests too.

Also have added the recommended deployment directory structure to be used by package manager. Let me know if this is ok.

Comment (pyohannes) [reply README.md:37]: We should add a _Requirements_ section, either here on in the `INSTALL.md`. I think the LightStep OpenTracing C++ project does a good job in describing requirements as well as detailed build instructions: https://github.com/lightstep/lightstep-tracer-cpp#requirements

File: README.md
Comment (pyohannes) [reply README.md:42]: It's just important that we have detailed CMake documentation somewhere, be it here on in the `INSTALL.md`. It should also cover the most useful build flags, like `WITH_OTLP`, `WITH_TESTS` and `WITH_EXAMPLES`.

File: RELEASING.md
Comment (lalitb) [reply RELEASING.md:7]: The steps to be followed as per this PR is
```
1. CHANGELOG.md is managed manually, and as part of raising any PR, the dev needs to add the summary of change under "Unreleased" section.
2. On day of release, Release manager decides on a tag say ( v0.0.1-beta).
3. He creates a branch ( pre_release_v0.0.1-beta) - this is done by running bash script.
4. He modified the CHANGELOG.md ( replacing Unreleased section with "<tag>-<date>", and adding empty "Unreleased" section on top for next release.) - Again, this is done by bash script
5. He creates a PR for these change log changes, and waits for it to be merged to master.
6.  Once PR is merged, he creates the above decided tag on the new merged commit hash.
7. He then goes to the github release UI page, and select "Create New Release". Specifies the already created tag in step 6, and takes the curated contents as release notes, and creates the release.
```
Please let me know what change we are expecting in these steps ? We can remove step 6 and let github release UI create the tag automatically, but wonder what will we achieve with that :)

File: README.md
Comment (pyohannes) [inline README.md:51]: I think this related to a previous discussion. However, I don't think we should put it here with this wording. We might need to document where `make install` puts our libraries, but we shouldn't recommend deployment structures for package managers.

Any proper DEB or RPM package will not install in `/usr/local`, but in `/usr/include` and `/usr/lib`. We should leave it to package maintainers to figure those things out (as I think we can't give valid general advice here).

File: pre_release.sh
Comment (lalitb) [reply pre_release.sh]: Thanks, Done :)

File: RELEASING.md
Comment (lalitb) [reply RELEASING.md:7]: > One imporant moment where we may need to get the tag of the repo is to generate the version.h from the tag, not auto-increment the tag itself.

I was thinking about this. Should we be having the release version in `version.h`. We do have `OPENTELEMETRY_VERSION` in `version.h` which seems to be used for dynamic loading of plugin libraries.

**Comment by lalitb** [comment]
> > let the tag be created automatically by github UI if that's the ask ?
> 
> Yeah, we can assume that Release Manager can use their best judgement and UI (no scripts) to generate the next tag, given UI totally allows to do this without running any command-line tooling ( or `gh` utility can be used by those who prefer command line tooling ).
> 
> I think it's more valuable to do the following: generate `version.h` from git tag. So that when the build is built - we properly report the actual SDK version. We can probably re-generate `version.h` in git commit hook for incremental builds that includes `commit id` (or manage this in CI). But if we build from source based on tag - the `version.h` should reflect that tag information. That is the build tooling or CMake or bazel should re-generate the local `version.h` file (or mod it, if stored in source).. SDK must be able to know (and even in most cases, report) its version at runtime.

OK Just saw this comment. Let me go through this now :)

File: README.md
Comment (lalitb) [reply README.md:51]: Thanks Johannes. I do somewhat agree with your comment.  @maxgolov  your thought : )

File: README.md
Comment (maxgolov) [inline README.md:48]: I like it that way. We can discuss more and tune.

**Review by maxgolov — APPROVED** [review]

**Comment by maxgolov** [comment]
>  Probably we can discuss this further in community meeting. I feel there are some gaps here.

We can describe the steps needed to create release / create tag with `gh`, and then publish the artifacts.... Maybe - as we publish the release using web-based UI, we can attach some form of a hook to automate the re-generation of the version header. Then even if Release Manager uses the release management tab / web-based UI to publish their notes, we can still properly patch the header.

In another project we, first:
- manually update the version header.
- check it in (thru PR review process).
- maybe at this point of PR review we should also review the release notes.
- then once the updated header is checked in -- we use web-based UX to create the tag + release notes under the release tab.

File: README.md
Comment (maxgolov) [reply README.md:51]: Ideally it should say `$INSTALL_PREFIX` instead of `/usr` - as this is either `/usr` or `/usr/local` . Maybe we should remove this and handle in a separate installation document, where we can describe the layout more precisely, depending on target OS and package manager: for rpm, deb, brew, vcpkg and others most common package managers. For CMake - we should also list what variables are populated by `find_package`, such as - include and library dirs of the opentelemetry package.

File: RELEASING.md
Comment (maxgolov) [reply RELEASING.md:7]: I think we absolutely must include the runtime detail (as in numeric and semver string version of the library) in the SDK header. This will be needed for exporters that must populate SDK version in the attributes, for analytics and for crash analysis / stability reasons.

Comment (maxgolov) [reply RELEASING.md:7]: There has to be two version concepts:
- version for API stability `VER_MAJOR.VER_MINOR`
- rest of semver 2.0 version for analytics and tracking; as there could be bugfix in SDK without a need to bump-up ver major and ver minor. 3rd tuple is the build increment, whereas the first two should be used to determine if the library has ABI-breaking changes.

Comment (maxgolov) [reply RELEASING.md:7]: e.g. no need to change the namespace version if no ABI breaking changes, but there is still a need to regenerate the build id information in the header, for runtime use.

File: README.md
Comment (pyohannes) [reply README.md:51]: > For CMake - we should also list what variables are populated by find_package, such as - include and library dirs of the opentelemetry package(s).

Do we intend to provide a CMake integration so the library can be discovered via `find_package`? We don't have this yet.

I'm fine with providing this directory list under the label of _default_ (CMake) locations, I'm a bit confused by labeling this as _recommended deployment directory structure_, as recommendations might vary considerable depending on the use case.

File: pre_release.sh
Comment (pyohannes) [inline pre_release.sh:65]: As far as I can see, this  doesn't add a new _Unreleased_ section, as claimed in the documentation.

File: README.md
Comment (pyohannes) [inline README.md:64]: Ideally, we'd have something like [this](https://github.com/open-telemetry/opentelemetry-java/blob/master/QUICKSTART.md) one day. But for now I think the reference to the example is fine.

File: README.md
Comment (pyohannes) [inline README.md:59]: Instead of TBD, can we leave this section out and create an issue to add it?

File: RELEASING.md
Comment (ThomsonTan) [reply RELEASING.md:7]: Could we also append the latest commit hash to the version number and embed it into the build binary? This will help troubleshooting for developers.

Comment (maxgolov) [reply RELEASING.md:7]: I think we can have both - short semver 2.0 in there with three tuples, but also some sort of `OsBuild` or `BUILD_ID`, which will be the full semver 2.0 --- with commit hash after dash. In many ways it's easier / nicer to work with just X.Y.Z, but agree for remote crash analysis - in prebuilt SDK shipped in products, we definitely need to capture and track the full build id including hash.

File: README.md
Comment (maxgolov) [reply README.md:51]: @pyohannes - I can work on this for CMake.

File: README.md
Comment (maxgolov) [reply README.md:59]: I think we should just skip the install entirely and handle it outside of README, splitting it maybe in separate docs:
- build with bazel
- build with cmake
- build with vcpkg (in essence, package manager helper around cmake)
- build with Visual Studio 2019 (CMake, GUI)
- easy-to-use one-liner build scripts in tools ( `build.sh` and `builld.cmd` for those who don't want to read the docs)

Comment (maxgolov) [reply README.md:59]: I'm actually working right now to do these write-ups..

File: pre_release.sh
Comment (lalitb) [reply pre_release.sh:65]: This [line](https://github.com/lalitb/opentelemetry-cpp/blob/2fcd4b3068a5374f6958842c11878106429abb10/pre_release.sh#L61) actually does something similar:
```
sed  -i "/\#\# \[Unreleased\]/a\\ \n\#\# \[${tag}\] ${date}"  $changelog_file
``` 
It searches for `##Unreleased` section and adds a new section with `<tag> - <date>`  header beneath it. That way `##Unreleased` section remains intact for new changes, and existing change-notes go within newly created section. I will reword the documentation.

Comment (lalitb) [reply pre_release.sh:65]: reworded the documentation,

File: README.md
Comment (lalitb) [reply README.md:59]: Have removed the TBD.

File: README.md
Comment (lalitb) [reply README.md:51]: Have removed all the install information, and just kept a reference to ./INSTALL.md. Hope that is fine.

**Comment by alolita** [comment]
Some items to consider for releasing artifacts:
* Add guidelines on commit comments
* Automate changelog generation
* Determine what features consititute major, minor, patch builds in line with semver.org guidelines
* Add PR builds, nightly builds and release builds
* Is there a need to have a compatibility matrix (e.g. ABI version compatibility)
Also to improve customer adoption:
* Make a sample app w the C++ library artifact in a Docker image available

File: RELEASING.md
Comment (jsuereth) [reply RELEASING.md:7]: QQ:  Does the tagged version ever make its way into a header file such that, e.g. we can use it in UserAgent strings when we write metrics?

**Review by jsuereth — APPROVED** [review]
under the assumption this is the *start* of the release process, looks great.

- I assume we'll be upadting the prepare_release script with more things we'd like.  Specifically I'd like runtime visibility into the release version.
- I like Max's suggestion of gh utilities for doing release.  The more automated the process can be the better.  However, I'd always prefer HAVING a process over not, so let's not let perfect be the enemy of good.

**Comment by reyang** [comment]
There are still many open comments/discussion here, @lalitb please follow up in separate PRs. Thanks!

File: RELEASING.md
Comment (maxgolov) [reply RELEASING.md:7]: > Does the tagged version ever make its way into a header file such that, e.g. we can use it in UserAgent strings when we write metrics?

@jsuereth - Absolutely! My opinion we must have it and it could be used as such. In some future OpenTelemetry exporters, that my team would be working on, it is absolutely necessary to identify the version of the instrumentation library in some dedicated field. For both - user engagement analytics and crash analytics. Generated version info as both string and numeric constant, is a good starting point for enabling that. We also can reuse that for DLLs, like - there is a way to include a C header in a resource file, to be embedded into DLL on Windows. So when you right-click on a DLL - you see precisely what version it is. Usually just `X.Y.Z` but some projects include the full SemVer 2.0 with commit ID.

Comment (lalitb) [reply RELEASING.md:7]: Yes. I have a git issue to add version information into header file.

**Comment by lalitb** [comment]
> Some items to consider for releasing artifacts:
> 
> * Add guidelines on commit comments
> * Automate changelog generation
> * Determine what features consititute major, minor, patch builds in line with semver.org guidelines
> * Add PR builds, nightly builds and release builds
> * Is there a need to have a compatibility matrix (e.g. ABI version compatibility)
>   Also to improve customer adoption:
> * Make a sample app w the C++ library artifact in a Docker image available

Really good points
 - This is incremental process, and most of these points are planned.

**Comment by lalitb** [comment]
> under the assumption this is the _start_ of the release process, looks great.
> 
> * I assume we'll be upadting the prepare_release script with more things we'd like.  Specifically I'd like runtime visibility into the release version.
> * I like Max's suggestion of gh utilities for doing release.  The more automated the process can be the better.  However, I'd always prefer HAVING a process over not, so let's not let perfect be the enemy of good.

Runtime visibility of release version is planned, we have an open ticket for this.

**Pull request merged**
